### ISSUE TITLE
IDispatchMessages and the casting of AddressTag

### ISSUE THREAD

**Issue by johnsimons** [Particular/NServiceBus#3214]
@SzymonPobiega I am upgrading RabbitMQ transport with @mikeminutillo and I came across the AddressTag polymorphic property.
It seems for implementers to use this they always have to cast its value to either Uni or Multi, at least that is what [MSMQ](https://github.com/Particular/NServiceBus/blob/a9def8b16c7ac3e428fb5747dbc553517b8505ed/src/NServiceBus.Core/Transports/Msmq/MsmqMessageSender.cs#L52) and [Sql](https://github.com/Particular/NServiceBus.SqlServer/blob/221016197e4ae4f102dbeb0ea44b2fbd8e05d8ae/src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs#L28) are doing

In the MSMQ implementation:
```c#
var routingStrategy = dispatchOptions.AddressTag as UnicastAddressTag;
```

I wonder if we should have a chat and see if we can make it more discoverable and natural, thougths ?

I was thinking of using two separate `Dispatch` methods, one for uni and another for multi, since Rabbit will use both. This way the contract is better expressed and there is no need for this casting or guessing what casting to use.
For MSMQ, you would just throw on the Multi.

So here is what I am proposing:
```c#
public interface IDispatchMessages
    {
        Task Dispatch(IEnumerable<TransportOperation> outgoingMessages, ContextBag context, UnicastAddressTag routing);

        Task Dispatch(IEnumerable<TransportOperation> outgoingMessages, ContextBag context, MulticastAddressTag routing);
    }
```

or the other option is to make the publishing an explicit concept at the transport level so:
```c#
public interface IDispatchMessages
    {
        Task Dispatch(IEnumerable<TransportOperation> outgoingMessages, ContextBag context, string destination);

        Task Publish(IEnumerable<TransportOperation> outgoingMessages, ContextBag context, Type messageType);
    }
```

@danielmarbach @SimonCropp @andreasohlund what do you think ?

**Comment by mikeminutillo** [comment]
I looked at this last week as well and couldn't find a way to implement it without casting the `AddressTag`. The code wasn't particularly clean. I like the first interface above and was investigating building that pattern into the transport internally. 

The base `AddressTag` abstract base class is empty and provides absolutely no value. The `AddressTag` property on the `DispatchOptions` may as well be of type `object`. At least with the interfaces above it could be polymorphically responsible for which of the two methods to call.

**Comment by danielmarbach** [comment]
Well if only we had pattern matching ;)

Really hesitant to bring the old way of dispatching back. On a dispatch
level core shouldn't care. What about a transport kit which does the
bridging to simplify transports? Could be source only?


-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

**Comment by mikeminutillo** [comment]
Will any transport ever be able to deal with AddressTag without updcasting it to something more concrete?

**Comment by SzymonPobiega** [comment]
@mikeminutillo there is no way to use it without the downcast. I am open to both options (downcast and two methods). No strong opinions about that.

**Comment by danielmarbach** [comment]
@SzymonPobiega Dispatch is dispatch. As soon as we introduce publish the semantics of that interface are no longer intention reveiling. I think we could easily start building a transport kit which takes a dependency on the core and provides some utils. Not everything has to be backed into the core.

**Comment by mikeminutillo** [comment]
@danielmarbach but we are already revealing that there are two types of addresses: Unicast and Multicast. Is it not reasonable to assume that they might be dispatched differently?

**Comment by johnsimons** [comment]
Sorry @danielmarbach, but as the API currently stands, it doesn't make much sense.
The fact that any implementation always have to downcast to use it is IMO a code smell.

**Comment by danielmarbach** [comment]
I think you fail to understand me and I fail to express myself so that I'm understandable. Happy to talk it through over voice.

**Comment by danielmarbach** [comment]
@mikeminutillo @johnsimons 

what about this:

```
    public interface IDispatchMessages
    {
        Task Dispatch(TransportOperations outgoingMessages, ContextBag context);
    }

    public class TransportOperations
    {
        IEnumerable<MultiCastTransportOperation> MulticastOperations { get; }

        IEnumerable<UnicastTransportOperation> UnicastOperations { get; }
    }

    public class MultiCastTransportOperation
    {

        public OutgoingMessage Message { get; private set; }

        public MulticastAddressTag AddressTag { get; private set; }

        public IEnumerable<DeliveryConstraint> DeliveryConstraints { get; private set; }

        public DispatchConsistency RequiredDispatchConsistency { get; set; }
    }

    public class UnicastTransportOperation
    {
        public OutgoingMessage Message { get; private set; }

        public MulticastAddressTag AddressTag { get; private set; }

        public IEnumerable<DeliveryConstraint> DeliveryConstraints { get; private set; }

        public DispatchConsistency RequiredDispatchConsistency { get; set; }
    }
```

This would the interface more strongly typed and also leave the dispatch operation as one single async call into the transport. The core would then make sure that it prepares the operations accordingly.

This would even allow us to implement proper sorting. If I remember correctly @SeanFeldman and @yvesgoeleven need to do some fancy sorting based on the address. We could also fulfill that by doing something like


```
    public class TransportOperationsFancy
    {
        IGrouping<MulticastAddressTag, IEnumerable<MultiCastTransportOperation>> MulticastOperations { get; }

        IGrouping<UnicastAddressTag, IEnumerable<UnicastTransportOperation>> UnicastOperations { get; }
    }
```

Thoughts?

cc @andreasohlund @SzymonPobiega

**Comment by andreasohlund** [comment]
I like it!

**Comment by johnsimons** [comment]
I think that works @danielmarbach.

But there is one more thing, why do we call it AddressTag ? Have we run out of ideas ?

**Comment by SeanFeldman** [comment]
@danielmarbach not sorting, but grouping, to be able to group commands and events based on destination and event type. It wasn't casting that was bad as much as not understanding from the API what should happen. The [snippets above](https://github.com/Particular/NServiceBus/issues/3214#issuecomment-164476321) do clarify that.

**Comment by danielmarbach** [comment]
@johnsimons In the brain of @szymonpobiega ;)

**Comment by johnsimons** [comment]
So @danielmarbach are you guys open to changing AddressTag to something else ?

**Comment by danielmarbach** [comment]
@johnsimons everything which expresses more intent makes sense for me. But I think we need to let @SzymonPobiega to chime in.

**Comment by SzymonPobiega** [comment]
`AddressTag` was the best thing we came up as part of 133 task force. The idea is that we stick it to the message to tell the dispatcher where to send it. If anybody has better ideas, feel free to propose them (PRs are always welcome). Nobody in the TF is attached emotionally to the names we gave to concepts.

On the other hand, I think that discussing these kinds of things for the sake of discussion is not really productive. I think the idea behind TFs was that they consist of pigs, people who do commit themselves to solving a particular problem. And if there are no *major* problems with the solution the TF came up, we should not have endless discussions on that.

To sum up, how about *PR or it didn't happen*?

**Comment by johnsimons** [comment]
@SzymonPobiega I am planning to submit a PR, I just wanted to make sure there wasn't something I missed before submitting a PR.
BTW, is this the right way to do it ?
I have seen lots of questions in slack nservicebus channel by @yvesgoeleven, is that the prefered way for questions ?
I am just afraid that questions in slack (especially at the aussie timezone) may go missing in the middle of everything else.

Finally, @SzymonPobiega I am not saying that the task force did a bad job at that time, I am just providing feedback on the current state of affairs, it hasn't been easy at all to re-write the RabbitMQ transport, there are a lot of new concepts (I guess majority of these concepts were always there just hidden away) to get familiar with, so please do not take this personally.

**Comment by timbussmann** [comment]
I'll work on this issue, starting with @danielmarbach's API proposal. Fully agree that we should avoid the need to cast `AddressTag`.

**Comment by johnsimons** [comment]
@timbussmann in a chat with @andreasohlund and @danielmarbach, we were also discussing whether makes sense to expose the Unicast/MulticastAddressTag classes to the implementers, they are just wrappers for the destination string and the event Type, so it may make sense for those to just be skipped and pass the values directly to the methods

**Comment by danielmarbach** [comment]
@johnsimons I think I made my standpoint clear. I like the strong typing of those value objects

**Comment by timbussmann** [comment]
well with `UnicastTransportOperation` and `MulticastTransportOperation` I don't see much value to have `AddressTag` classes from the transport perspective, but we can easily discuss that on the PR then?

**Comment by johnsimons** [comment]
Sorry @danielmarbach, I thought the strong typing was just for the internals, not the public api

**Comment by yvesgoeleven** [comment]
@johnsimons I wish I did not wake up with 200+ github notifications every morning, I can't keep up with unsubscribing myself from all those groups that I'm automatically added to. I missed this mention because of that, and therefore prefer to ask questions in slack so that I can get a (near) immediate response. I don't see 'messages getting lost' as a good argument, as I can't see the forest through the trees anyway.

On topic: I struggled with this change a lot, we do a lot of sorting indeed as ASB's batching support is subject to several constraints. First level sorting is not about Unicast/Multicast, it's about isolated/default dispatch, but it's conditional depending on the whether we are in a receive context or not, or wether the receive context was peeklock or receivedelete,etc.. Then move on to the next sorting steps, which include things like destination (in which unicast/multicast can have some influence), and batchsize, via-routing... 

Hence the split between unicast/multicast became a problem, I had to merge them back together again so that I could split by conistency first. But this turned out to not be so trivial as both operation classes share no common ground, so I had to copy all the data of these operations in my own types first, and then perform the sorting on our own metamodel.

So, long story short, ASB now has it's own metamodel representing Batch & BatchedOperation, which we'll use for sorting the batches going forward.

To make it easier for others, I do suggest however to give both operation classes some common ground like a shared interface or base class (Message, DeliveryConstraints, Consistency), so that it becomes possible to merge them again if different sorting is required.

**Comment by adamralph** [comment]
Is this a breaking API change?

**Comment by timbussmann** [comment]
@yvesgoeleven 
> To make it easier for others, I do suggest however to give both operation classes some common ground like a shared interface or base class (Message, DeliveryConstraints, Consistency), so that it becomes possible to merge them again if different sorting is required.

That should be easy to do. Does the Unicast/Multicast operation split make sense for all the other transports or do we need to rethink?

@adamralph do you mean the changes in this issue (yes) or the base class/interface (no) ?

**Comment by adamralph** [comment]
@timbussmann I mean any changes which are involved in solving this issue.

Given the answer is *yes*, this cannot be a refactoring, since a refactoring is a change to internal implementation details with no externally observable change.

It sound like we are introducing a new feature which includes a breaking change, would that be correct?

**Comment by danielmarbach** [comment]
Nitpicking: The feature was introduced when we added the dispatcher. This is a redesign of the dispatcher.

**Comment by adamralph** [comment]
I've reinstated the previous labels for now. I believe the following needs to done here, by the task force involved:

* Task force assigned and recorded in the issue description
* Consider what this issue represents, i.e. a feature, or something else?
* Re-title accorindgly
* Re-label accordingly - only one "type" label, any external change means it is not a refactoring
* Consider whether this is a breaking change and label if so

**Comment by yvesgoeleven** [comment]
@timbussmann hard to answer, it does not for ASB, ASQ will only care about the unicast ones, I see msmq and sql also only support unicast, their first concern is dispatch consistency as well. For Rabbit it seems to work out https://github.com/Particular/NServiceBus.RabbitMQ/blob/altMessagePump/src/NServiceBus.RabbitMQ/RabbitMqMessageSender.cs but it seems to ignore dispatch consistency all together

**Comment by timbussmann** [comment]
> ASB, ASQ will only care about the unicast ones, I see msmq and sql also only support unicast, their first concern is dispatch consistency as well. For Rabbit it seems to work out

that sounds like the changes from #3243 solved the need for casting but still has a weird abstraction to most transports. To me this sounds like `IDispatchMessages` could use a `DispatchIsolated` method.

Should we introduce the interface you mentioned to get you unstuck and raise another issue regarding the `IDispatchMessages` interface?

**Comment by danielmarbach** [comment]
I think for ASB we need the common base. After that we should wait until
RabbitMq properly implements it (sort by consistency...). My 2 cents

**Comment by timbussmann** [comment]
please review #3295

**Comment by johnsimons** [comment]
> After that we should wait until
> RabbitMq properly implements it (sort by consistency...)

Not following @danielmarbach, what is wrong with the way we are implementing the Rabbit ?

**Comment by timbussmann** [comment]
@yvesgoeleven with #3295 the base class should be available.

can we close this issue?

**Comment by yvesgoeleven** [comment]
It can be closed for me, but not sure if it can be closed for @danielmarbach 's comment about Rabbit

**Comment by danielmarbach** [comment]
Close
